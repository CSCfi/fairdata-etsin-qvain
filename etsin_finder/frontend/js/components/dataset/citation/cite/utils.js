import moment from 'moment'
import idnToLink from '../../../../utils/idnToLink'

const topOrg = org => {
  if (org.is_part_of) {
    return topOrg(org.is_part_of)
  }
  return org
}

// For surname formatting examples, see
// https://blog.apastyle.org/apastyle/2017/05/whats-in-a-name-two-part-surnames-in-apa-style.html

const startsWithLowerCase = str => str[0] === str[0].toLowerCase()

const suffixParts = ['Jr', 'Jr.', 'Sr', 'Sr.']

const lastnameParts = ['St', 'St.', 'Af', 'Av', 'De', 'Van', 'Von']

const romanNumeralMatch = /^(((IX|IV|V)I{0,3})|I{1,3})$/

const toInitial = name => `${name[0]}.`

export const getNameParts = name => {
  const parts = name.trim().split(' ')
  if (parts.length === 1) {
    return { first: [], last: [parts[0]], suffixes: [] }
  }

  const first = []
  const last = []
  const suffixes = []

  const lastPart = parts[parts.length - 1]
  if (suffixParts.includes(lastPart) || romanNumeralMatch.test(lastPart)) {
    suffixes.push(lastPart)
    parts.splice(parts.length - 1)
  }

  let isFirst = true
  parts.forEach((part, idx) => {
    if (idx === parts.length - 1 || startsWithLowerCase(part) || lastnameParts.includes(part)) {
      isFirst = false
    }
    if (isFirst) {
      first.push(part)
    } else {
      last.push(part)
    }
  })

  return {
    first,
    last,
    suffixes,
  }
}

export const getNameInitials = name => {
  const { first, last, suffixes } = getNameParts(name)

  const components = [last.join(' '), first.map(toInitial).join(' '), suffixes.join(' ')].filter(
    v => v
  )
  return components.join(', ')
}

export const getLastnameFirst = name => {
  const { first, last, suffixes } = getNameParts(name)

  const components = [last.join(' '), first.join(' '), suffixes.join(' ')].filter(v => v)
  return components.join(', ')
}

export const getAuthorsInitial = (dataset, getTranslation, etAlThreshold) => {
  const authors = (dataset.research_dataset.creator || []).map(author => {
    if (author['@type'] === 'Organization') {
      return getTranslation(topOrg(author)?.name)
    }
    return getNameInitials(getTranslation(author.name))
  })
  if (authors.length > etAlThreshold) {
    authors.splice(etAlThreshold - 1, authors.length - etAlThreshold)
    authors[authors.length - 1] = `. . . ${authors[authors.length - 1]}`
  } else if (authors.length > 1) {
    authors[authors.length - 1] = `& ${authors[authors.length - 1]}`
  }
  return authors
}

export const getAuthorsFull = (dataset, getTranslation, etAlThreshold, etAlCount) => {
  const authors = (dataset.research_dataset.creator || []).map((author, index) => {
    if (author['@type'] === 'Organization') {
      return getTranslation(topOrg(author)?.name)
    }
    if (index === 0) {
      return getLastnameFirst(getTranslation(author.name))
    }
    return getTranslation(author.name)
  })
  if (authors.length > etAlThreshold) {
    authors.splice(etAlCount)
    authors[authors.length - 1] = `${authors[authors.length - 1]}, et al.`
  } else if (authors.length > 1) {
    authors[authors.length - 1] = `and ${authors[authors.length - 1]}`
  }
  return authors
}

export const addParens = string => {
  if (string === undefined) {
    return undefined
  }
  return `(${string})`
}

export const addQuotes = string => {
  if (string === undefined) {
    return undefined
  }
  return `”${string}”`
}

export const capitalizeFirst = string => {
  if (string === undefined) {
    return undefined
  }
  return `${string.charAt(0).toUpperCase()}${string.slice(1)}`
}

export const getInitial = string => {
  if (string === undefined) {
    return undefined
  }
  return `${string.charAt(0).toUpperCase()}.`
}

export const getYear = dataset => {
  const issued = dataset.research_dataset.issued
  if (!issued) {
    return undefined
  }
  const year = moment(issued).format('YYYY')
  return year
}

export const getTitle = (dataset, getTranslation) => {
  if (!dataset.research_dataset.title) {
    return undefined
  }
  return getTranslation(dataset.research_dataset.title)
}

export const getVersion = (dataset, getTranslation) => {
  if (!dataset.dataset_version_set) {
    return undefined
  }
  const identifier = dataset.identifier
  const versionIndex = dataset.dataset_version_set.findIndex(v => v.identifier === identifier)
  if (versionIndex < 0) {
    return undefined
  }
  const version = dataset.dataset_version_set.length - versionIndex
  return getTranslation({ en: `Version ${version}`, fi: `versio ${version}` })
}

export const getIdentifier = dataset => {
  const identifier = dataset.research_dataset.preferred_identifier
  if (!identifier) {
    return undefined
  }
  const url = idnToLink(identifier)
  if (url) {
    return url
  }
  if (dataset.draft_of) {
    return dataset.draft_of.preferred_identifier
  }
  if (dataset.state === 'draft') {
    return undefined
  }
  return identifier
}

export const getPublisher = (dataset, getTranslation) => {
  const pub = dataset.research_dataset.publisher
  if (pub) {
    const top = topOrg(pub)
    if (top !== pub) {
      return `${getTranslation(top.name)}, ${getTranslation(pub.name)}`
    }
    return getTranslation(pub.name)
  }
  return undefined
}
