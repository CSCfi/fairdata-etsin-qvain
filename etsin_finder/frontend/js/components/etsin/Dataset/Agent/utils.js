/* eslint-disable camelcase */
export const hasExtraInfo = agent =>
  agent.person || 
  agent.contributor_role || 
  agent.contributor_type || 
  agent.organization.parent || 
  agent.organization.homepage || 
  agent.organization.url

export const flatParentOrgs = (agent) => {
  let org = agent.organization
  const orgs = []
  while (org) {
    orgs.unshift({ ...org })
    org = org.parent
  }
  return orgs
}

export const getOrgKey = org => org.url || Object.values(org.pref_label)[0]

export const getRefDataKey = data => data.identifier || Object.values(data.pref_label)[0]
