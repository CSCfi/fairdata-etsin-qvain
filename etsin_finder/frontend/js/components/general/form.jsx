{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}

import styled from 'styled-components'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faInfoCircle } from '@fortawesome/free-solid-svg-icons'
import withCustomProps from '@/utils/withCustomProps'

export const HelpIconStyles = styled(FontAwesomeIcon)`
  &:hover {
    color: ${props => props.theme.color.primary};
  }
`

export const NoStyleButton = withCustomProps(styled.button)`
  border: none;
  background-color: unset;
  padding-left: 0;
`

export const HelpIcon = props => (
  <NoStyleButton {...props} type="button">
    <HelpIconStyles icon={faInfoCircle} />
  </NoStyleButton>
)
