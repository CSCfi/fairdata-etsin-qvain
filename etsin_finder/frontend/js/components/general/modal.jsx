{
  /**
   * This file is part of the Etsin service
   *
   * Copyright 2017-2018 Ministry of Education and Culture, Finland
   *
   *
   * @author    CSC - IT Center for Science Ltd., Espoo Finland <servicedesk@csc.fi>
   * @license   MIT
   */
}

import { useState } from 'react'
import styled from 'styled-components'
import ReactModal from 'react-modal'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faTimes } from '@fortawesome/free-solid-svg-icons'
import Translate from '@/utils/Translate'

import ErrorBoundary from './errorBoundary'
import { TransparentButton } from './button'

const defaultStyles = {
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.15)',
    zIndex: 2,
  },
  content: {
    top: '0',
    bottom: '0',
    left: '0',
    right: '0',
    position: 'relative',
    maxHeight: '80vh',
    minWidth: '30vw',
    maxWidth: '600px',
    margin: '0.5em',
    border: 'none',
    padding: '2em',
    boxShadow: '0px 6px 12px -3px rgba(0, 0, 0, 0.15)',
  },
}

if (BUILD !== 'test') ReactModal.setAppElement('#root')

const Modal = ({
  isOpen,
  contentLabel,
  children,
  labelledBy,
  onAfterOpen = () => {},
  onRequestClose = () => {},
  customStyles = {},
}) => {
  const [clearError, setClearError] = useState()

  const getStyles = () => {
    if (clearError) {
      return defaultStyles
    }

    const overlayStyle = { ...defaultStyles.overlay, ...customStyles.overlay }
    const contentStyle = { ...defaultStyles.content, ...customStyles.content }
    return { overlay: overlayStyle, content: contentStyle }
  }

  const errorCallback = (_error, _info, clearError) => {
    setClearError(clearError)
  }

  const requestClose = () => {
    if (clearError) {
      clearError()
    }
    onRequestClose()
  }

  return (
    <ReactModal
      isOpen={isOpen}
      onAfterOpen={onAfterOpen}
      onRequestClose={requestClose}
      style={getStyles()}
      contentLabel={contentLabel}
      aria={{ labelledby: labelledBy }}
    >
      <Translate
        component={CloseButton}
        onClick={onRequestClose}
        noMargin
        attributes={{ 'aria-label': 'qvain.common.close' }}
      >
        <FontAwesomeIcon aria-hidden icon={faTimes} />
      </Translate>
      <ErrorBoundary callback={errorCallback}>{children}</ErrorBoundary>
    </ReactModal>
  )
}

Modal.propTypes = {
  customStyles: PropTypes.object,
  isOpen: PropTypes.bool.isRequired,
  onAfterOpen: PropTypes.func,
  onRequestClose: PropTypes.func,
  contentLabel: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  labelledBy: PropTypes.string,
}

const CloseButton = styled(TransparentButton)`
  position: absolute;
  top: 1em;
  right: 1em;
`

export default Modal
