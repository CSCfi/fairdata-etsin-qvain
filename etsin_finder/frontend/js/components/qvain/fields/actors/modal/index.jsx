import React, { useState } from 'react'
import { observer } from 'mobx-react'
import Translate from 'react-translate-component'
import styled from 'styled-components'

import ActorTypeSelect from './actorTypeSelect'
import ActorRoles from './actorRoles'
import ActorInfo from './actorInfo'
import Modal from '../../../../general/modal'
import { ActorIcon } from '../common'
import { actorSchema } from '../../../utils/formValidation'
import { ConfirmClose } from '../../../general/modal/confirmClose'
import ActorErrors from './actorErrors'
import Buttons from './buttons'
import { TableButton } from '../../../general/buttons'
import { useStores } from '../../../utils/stores'

export const ActorModalBase = () => {
  const {
    Qvain: {
      Actors: {
        actorInEdit,
        editActor,
        saveActor,
        cancelActor,
        actors,
        referenceOrganizationErrors,
      },
      readonly,
    },
  } = useStores()
  const [actorError, setActorError] = useState(null)
  const [confirmClose, setConfirmClose] = useState(false)

  if (!actorInEdit) {
    return null
  }

  const loadingFailed = Object.values(referenceOrganizationErrors).length > 0
  const originalActor = actors.find(actor => actor.uiid === actorInEdit.uiid)
  const isNew = !originalActor
  const action = isNew ? 'create' : 'edit'

  const handleSaveActor = () => {
    actorSchema
      .validate(actorInEdit, { strict: true })
      .then(() => {
        saveActor(actorInEdit)
        editActor(null)
        setActorError(null)
      })
      .catch(err => {
        setActorError(err.errors)
      })
  }

  const close = () => {
    setConfirmClose(false)
    setActorError(null)
    cancelActor()
  }

  const handleRequestClose = hasChanged => {
    if (hasChanged) {
      setConfirmClose(true)
    } else {
      close()
    }
  }

  const hideConfirmClose = () => {
    setConfirmClose(false)
  }

  const hasChanged = () => {
    if (originalActor) {
      return JSON.stringify(originalActor) !== JSON.stringify(actorInEdit)
    }
    return (
      Object.values(actorInEdit.person.name).some(val => val.length > 0) ||
      actorInEdit.person.email.length ||
      actorInEdit.person.identifier.length ||
      actorInEdit.roles.length ||
      actorInEdit.organizations.length
    )
  }

  const requestClose = () => handleRequestClose(hasChanged())

  return (
    <Modal
      isOpen
      onRequestClose={requestClose}
      contentLabel="fixDeprecatedModal"
      customStyles={modalStyle}
    >
      <Header>
        <ActorIcon actor={actorInEdit} style={{ marginRight: '1rem' }} />
        <Translate content={`qvain.actors.add.action.${action}`} />
      </Header>
      <Content>
        <ActorTypeSelect />
        <ActorRoles />
        <ActorInfo />
      </Content>
      <ActorErrors actorError={actorError} loadingFailed={loadingFailed} />
      <Buttons
        handleRequestClose={requestClose}
        handleSaveActor={handleSaveActor}
        readonly={readonly}
      />

      <ConfirmClose show={confirmClose} onCancel={hideConfirmClose} onConfirm={close} />
    </Modal>
  )
}

export const AutoWidthTableButton = styled(TableButton)`
  width: auto;
  margin-bottom: 0.5rem;
`

const Content = styled.div`
  flex-grow: 1;
  width: 100%;
  overflow: auto;
  margin-bottom: 0.5rem;
`

const Header = styled.h3`
  margin-right: 1.5rem;
  margin-top: -0.5rem;
  margin-bottom: 0.5rem;
  margin-left: 0rem;
`

const modalStyle = {
  content: {
    top: '0',
    bottom: '0',
    left: '0',
    right: '0',
    position: 'relative',
    minHeight: '85vh',
    maxHeight: '95vh',
    minWidth: '300px',
    maxWidth: '600px',
    margin: '0.5em',
    border: 'none',
    padding: '2em',
    boxShadow: '0px 6px 12px -3px rgba(0, 0, 0, 0.15)',
    overflow: 'hidden',
    paddingLeft: '2em',
    paddingRight: '2em',
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
  },
}

export default observer(ActorModalBase)
