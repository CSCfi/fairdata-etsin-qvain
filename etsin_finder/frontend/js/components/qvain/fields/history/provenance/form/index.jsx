import React, { useState } from 'react'
import PropTypes from 'prop-types'
import { FormContainer } from '../../../../general/modal/form'
import TabInput from '../../../../general/input/translationTabInputModal'
import TranslationTab from '../../../../general/input/translationTab'
import LocationInput from './locationInput'
import ModalReferenceInput from '../../../../general/modal/modalReferenceInput'
import { Outcome, Lifecycle } from '../../../../../../stores/view/qvain/qvain.provenances'
import Separator from '../../../../general/modal/modalSeparator'
import ActorsInput from './actorsInput'
import DurationPicker from '../../../../general/input/durationpicker'
import { useStores } from '../../../../utils/stores'

const Form = ({ Field }) => {
  const {
    Locale: { getMatchingLang },
  } = useStores()
  const { inEdit } = Field

  const translations = [inEdit.name, inEdit.description, inEdit.outcomeDescription]
  const [language, setLanguage] = useState(getMatchingLang(translations))

  return (
    <FormContainer>
      <TranslationTab language={language} setLanguage={setLanguage}>
        <TabInput Field={Field} datum="name" language={language} isRequired />
        <TabInput Field={Field} datum="description" language={language} />
        <TabInput Field={Field} datum="outcomeDescription" language={language} />
      </TranslationTab>
      <Separator />
      <DurationPicker
        Field={Field}
        datum="periodOfTime"
        language={language}
        id="provenance-period"
      />
      <LocationInput />
      <Separator />
      <ModalReferenceInput
        Field={Field}
        datum="outcome"
        metaxIdentifier="event_outcome"
        model={Outcome}
      />
      <Separator />
      <ActorsInput Field={Field} datum="actors" />
      <Separator />
      <ModalReferenceInput
        Field={Field}
        datum="lifecycle"
        metaxIdentifier="lifecycle_event"
        model={Lifecycle}
      />
    </FormContainer>
  )
}

Form.propTypes = {
  Field: PropTypes.object.isRequired,
}

export default Form
