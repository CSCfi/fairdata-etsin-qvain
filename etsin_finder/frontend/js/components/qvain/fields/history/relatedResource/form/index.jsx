import React, { useState } from 'react'
import PropTypes from 'prop-types'
import TranslationTab from '../../../../general/input/translationTab'
import TabInput from '../../../../general/input/translationTabInputModal'
import { FormContainer } from '../../../../general/modal/form'
import ModalInput from '../../../../general/modal/modalInput'
import ModalReferenceInput from '../../../../general/modal/modalReferenceInput'
import { RelationType } from '../../../../../../stores/view/qvain/qvain.relatedResources'
import ModalSeparator from '../../../../general/modal/modalSeparator'
import { useStores } from '../../../../utils/stores'

const Form = ({ Field, hideRelationType }) => {
  const {
    Locale: { getMatchingLang },
  } = useStores()
  const { inEdit } = Field

  const translations = [inEdit.name, inEdit.description]
  const [language, setLanguage] = useState(getMatchingLang(translations))

  return (
    <FormContainer>
      <TranslationTab language={language} setLanguage={setLanguage}>
        <TabInput Field={Field} datum="name" language={language} isRequired />
        <TabInput Field={Field} datum="description" language={language} />
      </TranslationTab>
      <ModalInput Field={Field} datum="identifier" />
      <ModalReferenceInput
        Field={Field}
        datum="entityType"
        metaxIdentifier="resource_type"
        model={RelationType}
      />
      <ModalSeparator />
      {!hideRelationType && (
        <ModalReferenceInput
          Field={Field}
          datum="relationType"
          metaxIdentifier="relation_type"
          model={RelationType}
          isRequired
        />
      )}
    </FormContainer>
  )
}

Form.propTypes = {
  Field: PropTypes.object.isRequired,
  hideRelationType: PropTypes.bool,
}

Form.defaultProps = {
  hideRelationType: false,
}

export default Form
