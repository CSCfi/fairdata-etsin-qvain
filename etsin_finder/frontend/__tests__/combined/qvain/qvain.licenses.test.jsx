import Harness from '../../components/componentTestHarness'
import React, { useState } from 'react'
import { shallow } from 'enzyme'
import Translate from 'react-translate-component'
import { components } from 'react-select'
import CreatableSelect from 'react-select/creatable'

import etsinTheme from '../../../js/styles/theme'
import { buildStores } from '../../../js/stores'
import { useStores } from '../../../js/stores/stores'
import RightsAndLicenses from '../../../js/components/qvain/fields/licenses'
import LicensesComponent, { License } from '../../../js/components/qvain/fields/licenses/licenses'
import AccessTypeComponent, {
  AccessType,
} from '../../../js/components/qvain/fields/licenses/accessType'
import RestrictionGrounds from '../../../js/components/qvain/fields/licenses/restrictionGrounds'
import { ACCESS_TYPE_URL, LICENSE_URL } from '../../../js/utils/constants'
import { ValidationErrors } from '../../../js/components/qvain/general/errors/validationError'
import Tooltip from '../../../js/components/qvain/general/section/tooltip'
import { HelpIcon } from '../../../js/components/qvain/general/modal/form'
import Licenses from '../../../js/stores/view/qvain/qvain.license'
import LicensesInfo from '../../../js/components/qvain/fields/licenses/licensesInfo'

jest.mock('../../../js/stores/stores', () => {
  const mockUseStores = jest.fn()
  return {
    ...jest.requireActual('../../../js/stores/stores'),
    useStores: mockUseStores,
  }
})

jest.mock('react', () => {
  const original = jest.requireActual('react')
  return {
    ...original,
    useState: jest.fn(original.useState),
  }
})

describe('Qvain.RightsAndLicenses', () => {
  let stores

  const harness = new Harness(RightsAndLicenses)

  const getRenderedLicenseUrls = shallowLicenseComponent => {
    const selectedOptions = shallowLicenseComponent
      .findWhere(c => c.prop('component') === CreatableSelect)
      .dive()
      .dive()
      .dive()
      .dive()
      .find(components.MultiValue)
    return selectedOptions.map(opt => opt.prop('data').identifier)
  }

  beforeEach(() => {
    stores = buildStores()
    stores.Qvain.resetQvainStore()
    useStores.mockReturnValue(stores)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render default license', () => {
    const component = shallow(<License Stores={stores} theme={etsinTheme} />)
    expect(getRenderedLicenseUrls(component)).toEqual([LICENSE_URL.CCBY4])
  })

  it('should render one added license, Other (URL)', () => {
    const { set: setLicenseArray, Model: LicenseConstructor } = stores.Qvain.Licenses
    setLicenseArray([
      LicenseConstructor({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'https://test.url'),
    ])
    const component = shallow(<License Stores={stores} theme={etsinTheme} />)
    expect(getRenderedLicenseUrls(component)).toEqual(['https://test.url'])
  })

  it('should render one added license, CCBY4', () => {
    const { set: setLicenseArray, Model: LicenseConstructor } = stores.Qvain.Licenses
    setLicenseArray([])
    setLicenseArray([
      LicenseConstructor(
        { en: 'Creative Commons Attribution 4.0 International (CC BY 4.0)' },
        LICENSE_URL.CCBY4
      ),
    ])
    const component = shallow(<License Stores={stores} theme={etsinTheme} />)
    expect(getRenderedLicenseUrls(component)).toEqual([LICENSE_URL.CCBY4])
  })

  it('should render three added licenses, Other (URL) x 2 + CCBY4', () => {
    const { set: setLicenseArray, Model: LicenseConstructor } = stores.Qvain.Licenses
    setLicenseArray([
      LicenseConstructor({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'https://test.url'),
      LicenseConstructor({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'https://test2.url'),
      LicenseConstructor(
        { en: 'Creative Commons Attribution 4.0 International (CC BY 4.0)' },
        LICENSE_URL.CCBY4
      ),
    ])
    const component = shallow(<License Stores={stores} theme={etsinTheme} />)
    expect(getRenderedLicenseUrls(component)).toEqual([
      'https://test.url',
      'https://test2.url',
      LICENSE_URL.CCBY4,
    ])
  })

  it('should render four licenses where two have errors', () => {
    const { Licenses } = stores.Qvain
    Licenses.set([
      Licenses.Model(
        { en: 'Creative Commons Attribution 4.0 International (CC BY 4.0)' },
        LICENSE_URL.CCBY4
      ),
      Licenses.Model({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'httpöötest.url'),
      Licenses.Model({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'http://ok.url'),
      Licenses.Model({ en: 'Other (URL)', fi: 'Muu (URL)' }, 'httppp:/fail.url'),
    ])
    const component = shallow(<License Stores={stores} theme={etsinTheme} />)
    component.instance().validateLicenses()
    expect(getRenderedLicenseUrls(component)).toEqual([
      LICENSE_URL.CCBY4,
      'httpöötest.url',
      'http://ok.url',
      'httppp:/fail.url',
    ])
    const errors = component
      .find(ValidationErrors)
      .dive()
      .find(Translate) // ValidationErrorItem
      .map(item => item.text())

    expect(errors.length).toBe(2)
    expect(errors[0].startsWith('httpöötest.url'))
    expect(errors[1].startsWith('httppp:/fail.url'))
  })

  it('should render <AccessType />', () => {
    const component = shallow(<AccessType Stores={stores} />)
    expect(component).toMatchSnapshot()
  })

  it('should render <RestrictionGrounds />', () => {
    const { set: setAccessType, Model: AccessTypeConstructor } = stores.Qvain.AccessType
    setAccessType(AccessTypeConstructor(undefined, ACCESS_TYPE_URL.EMBARGO))
    const component = shallow(<AccessType Stores={stores} />)
    expect(component.find(RestrictionGrounds).length).toBe(1)
  })

  it('should NOT render <RestrictionGrounds />', () => {
    const { set: setAccessType, Model: AccessTypeConstructor } = stores.Qvain.AccessType
    setAccessType(AccessTypeConstructor(undefined, ACCESS_TYPE_URL.OPEN))
    const component = shallow(<AccessType Stores={stores} />)
    expect(component.find(RestrictionGrounds).length).toBe(0)
  })

  describe('render', () => {
    beforeEach(() => {
      harness.shallow()
    })

    test('should exist', () => {
      harness.shouldExist()
    })

    test('should have children with expedted properties', () => {
      const children = [
        {
          label: 'Title',
          findType: 'prop',
          findArgs: ['content', 'qvain.rightsAndLicenses.title'],
        },
        {
          label: 'Tooltip',
          findArgs: Tooltip,
        },
        {
          label: 'Help',
          findArgs: HelpIcon,
        },
        {
          label: 'Licenses',
          findArgs: LicensesComponent,
        },
        {
          label: 'AccessType',
          findArgs: AccessTypeComponent,
        },
      ]

      const props = {
        Tooltip: {
          isOpen: false,
          align: 'Right',
          text: <LicensesInfo />,
        },
      }

      harness.shouldIncludeChildren(children, props)
    })

    describe('given tootlipOpen: true, when Tooltip close', () => {
      const setTooltipOpen = jest.fn()
      let tooltipOpen = true

      beforeEach(() => {
        useState.mockImplementationOnce(() => [tooltipOpen, setTooltipOpen])
        harness.shallow()
        harness.find(Tooltip)
        harness.props.close()
      })

      test('should call setTooltipOpen false', () => {
        expect(setTooltipOpen).toHaveBeenCalledWith(false)
      })
    })

    describe('given tooltipOpen: false, when Tooltip close', () => {
      const setTooltipOpen = jest.fn()
      let tooltipOpen = false

      beforeEach(() => {
        useState.mockImplementationOnce(() => [tooltipOpen, setTooltipOpen])
        harness.shallow()
        harness.find(Tooltip)
        harness.props.close()
      })

      test('should call setTooltipOpen with true', () => {
        expect(setTooltipOpen).toHaveBeenCalledWith(true)
      })
    })

    describe('given tootlipOpen: true, when HelpIcon onClick', () => {
      const setTooltipOpen = jest.fn()
      let tooltipOpen = true

      beforeEach(() => {
        useState.mockImplementationOnce(() => [tooltipOpen, setTooltipOpen])
        harness.shallow()
        harness.find(HelpIcon)
        harness.trigger('click')
      })

      test('should call setTooltipOpen false', () => {
        expect(setTooltipOpen).toHaveBeenCalledWith(false)
      })
    })

    describe('given tooltipOpen: false, when HelpIcon onClick', () => {
      const setTooltipOpen = jest.fn()
      let tooltipOpen = false

      beforeEach(() => {
        useState.mockImplementationOnce(() => [tooltipOpen, setTooltipOpen])
        harness.shallow()
        harness.find(HelpIcon)
        harness.trigger('click')
      })

      test('should call setTooltipOpen with true', () => {
        expect(setTooltipOpen).toHaveBeenCalledWith(true)
      })
    })
  })
})
