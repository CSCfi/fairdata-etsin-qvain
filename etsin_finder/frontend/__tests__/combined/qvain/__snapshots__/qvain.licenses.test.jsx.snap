// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Qvain.RightsAndLicenses should render <AccessType /> 1`] = `
<Card
  bottomContent={false}
>
  <form__LabelLarge
    htmlFor="accessTypeSelect"
  >
    <Translate
      component="span"
      content="qvain.rightsAndLicenses.accessType.title"
    />
  </form__LabelLarge>
  <Translate
    attributes={
      Object {
        "placeholder": "qvain.rightsAndLicenses.accessType.placeholder",
      }
    }
    clearable={true}
    component={[Function]}
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    inputId="accessTypeSelect"
    isDisabled={false}
    name="accessType"
    onBlur={[Function]}
    onChange={[Function]}
    options={Array []}
    value={
      Object {
        "name": undefined,
        "url": "http://uri.suomi.fi/codelist/fairdata/access_type/code/open",
      }
    }
  />
  <ValidationError />
</Card>
`;
